INT-4365: Improve notPropagatedHeaders function

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4365

It is much useful to configure the `notPropagatedHeaders` as a set of
patterns to match.
In this case we can filter a group of headers with the common prefix or
suffix

* Allow to configure `AbstractMessageProducingHandler.setNotPropagatedHeaders`
as simple patterns; the `*` means filter all - not copy request headers
at all - similar to `transformer` behavior
* Add `ConsumerEndpointSpec.notPropagatedHeaders()` for Java DSL
* Add `not-propagated-headers` to the `<service-activator>`

artembilan committed:
INT-4365: Improve notPropagatedHeaders function

JIRA: https://jira.spring.io/browse/INT-4365

It is much useful to configure the `notPropagatedHeaders` as a set of
patterns to match.
In this case we can filter a group of headers with the common prefix or
suffix

* Allow to configure `AbstractMessageProducingHandler.setNotPropagatedHeaders`
as simple patterns; the `*` means filter all - not copy request headers
at all - similar to `transformer` behavior
* Add `ConsumerEndpointSpec.notPropagatedHeaders()` for Java DSL
* Add `not-propagated-headers` to the `<service-activator>`

andybaer commented on src/reference/asciidoc/message.adoc:
```diff
@@ -394,9 +394,16 @@ Also, a header is only propagated if it does not already exist in the outbound m
-IMPORTANT: Header propagation suppression does not apply to those endpoints that don't modify the message, e.g. <<bridge, bridges>> and <<router, routers>>
+Starting with _version 5.0_, the `setNotPropagatedHeaders()` implementation on the `AbstractMessageProducingHandler` applies the simple patterns (`xxx*`, `*xxx`, `*xxx*` or `xxx*yyy`) to allow to filter header with common suffix or prefix.
+See `PatternMatchUtils` JavaDocs for more information.
+When `*` (asterisk) is used, no any headers are propagated.
```
When `*` (asterisk) is used, no headers are propagated.

m-terra reviewed (commented):
Does this PR included requires-reply xml support for Transformer?
These changes will give us the flexibility and control we need in our complex flows. 
Thanks!

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/AbstractMessageProducingHandler.java:
```diff
@@ -134,22 +140,26 @@ private void updateNotPropagatedHeaders(String[] headers, boolean merge) {
 		}
 		this.selectiveHeaderPropagation = this.notPropagatedHeaders.size() > 0;
+
+		this.shouldCopyRequestHeaders = !this.notPropagatedHeaders.contains("*");
```
That's not. Because `notPropagatedHeaders` is a `Set` not simple `String`. So, we don't copy headers at all if one of the `notPropagatedHeaders` elements is exactly asterisk, e.g. `foo, b*, *, b*z`.
So, all other patterns will be ignored just because we have one of them as `*`. It has a highest priority.

andybaer replied:
Of course, sry!

artembilan committed:
Address PR comments; some other improvements

* Fix `ConsumerEndpointSpec#notPropagatedHeaders()` log message
* Improve `AbstractMessageProducingHandler.notPropagatedHeaders() logic
so any `*` in the set of patterns eliminates all others since it has
a highest priority
* Expose `requires-reply` for the `<transformer>` as a `true` by default
* Refactor a bit `AbstractStandardMessageHandlerFactoryBean` hierarchy to
avoid duplicated code

artembilan commented:
Thank you, @andybaer , for the valuable fieedback!

Well, I did requested changes - please take a look again.

Right, I was in doubt last week to expose `requiresReply` for the `<transformer>` or not. 
That is not the component to return a null as simple result of the transformation function.
Although that's for now... I still hope that we will introduce once `null payload` support in the future: https://jira.spring.io/browse/INT-997

But after some weekend debates in my head I decided to relax a restriction and expose it as is.
In the end that doesn't hurt the Framework and everything rest is up to the end-user application. We provide an option and you decide to use it or not and how.

andybaer commented:
Thanks Artem!
For our application, it would be either 
- using transformer everywhere and mark possible 'flow aborters' with requires-reply=false
or
- using service activators everywhere and mark possible incorrect header merges with not-progated-headers=*
For readability I would prefer the first one but most important is having an option to deal with the scenarios.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/handler/AbstractMessageProducingHandler.java:
```diff
@@ -133,23 +139,34 @@ private void updateNotPropagatedHeaders(String[] headers, boolean merge) {
+		if (hasAsterisk) {
+			this.notPropagatedHeaders.clear();
+			this.notPropagatedHeaders.add("*");
+		}
```
I would just set a boolean in this case so we can optimize later.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/handler/AbstractMessageProducingHandler.java:
```diff
@@ -355,12 +372,15 @@ else if (output instanceof AbstractIntegrationMessageBuilder) {
 			builder = this.getMessageBuilderFactory().withPayload(output);
 		}
-		if (this.shouldCopyRequestHeaders()) {
+		if (shouldCopyRequestHeaders()) {
```
If (shouldCopyRequestHeaders() && !this.copyNoHeaders) {...

artembilan replied:
I don't see reason in that `copyNoHeaders` `shouldCopyRequestHeaders` does the trick.

artembilan replied:
Remember:
> if * {@link #shouldCopyRequestHeaders() shouldCopyRequestHeaaders} is true. 

?

So, independently of what you configure for the `headers`, but if `shouldCopyRequestHeaders()` returns `false`, they all are ignored.

garyrussell replied:
But what if a subclass overrides `shouldCopyRequestHeaders()` to return `true` and `*` is the only pattern?

artembilan replied:
Ha! That was your suggestion to go pattern way 😄 
I was just suggested simple boolean flag - for symmetry of existing `shouldCopyRequestHeaders()`.

I don't know (yet) how to be...

Need to think, but I somehow don't like too much boolean flags...

garyrussell replied:
Perhaps we could call `shouldCopyRequestHeaders()` just once instead of on every message. It doesn't really make much sense to call it on every message anyway (unless a subclass used a `ThreadLocal`).

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/handler/AbstractMessageProducingHandler.java:
```diff
@@ -133,23 +139,34 @@ private void updateNotPropagatedHeaders(String[] headers, boolean merge) {
+		if (hasAsterisk) {
+			this.notPropagatedHeaders.clear();
+			this.notPropagatedHeaders.add("*");
+		}
```
I would just set a boolean in this case so we can optimize later e.g `this.copyNoHeaders=true`.

artembilan replied:
I do that. See below:
```
this.shouldCopyRequestHeaders = !hasAsterisk;
```
And `shouldCopyRequestHeaders()` body.
What am I missing?

garyrussell replied:
Ah - didn't see that but, since subclasses can (and do) override that method, we need another boolean.

artembilan replied:
Right, and they always override it to `false` eliminating all the purpose of this `notPropagatedHeaders`

garyrussell replied:
While ours might override to `false`, we can't guarantee user handlers do the same.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/handler/AbstractMessageProducingHandler.java:
```diff
@@ -355,12 +372,15 @@ else if (output instanceof AbstractIntegrationMessageBuilder) {
-				for (String header : this.notPropagatedHeaders) {
-					headersToCopy.remove(header);
-				}
+				String[] patterns = this.notPropagatedHeaders.toArray(new String[this.notPropagatedHeaders.size()]);
```
We can optimize here rather than building an array for every message.

garyrussell commented on src/reference/asciidoc/message.adoc:
```diff
@@ -394,9 +394,16 @@ Also, a header is only propagated if it does not already exist in the outbound m
 You can also globally suppress propagation of specific message headers by setting the `readOnlyHeaders` property in `META-INF/spring.integration.properties` to a comma-delimited list of headers.
 
-IMPORTANT: Header propagation suppression does not apply to those endpoints that don't modify the message, e.g. <<bridge, bridges>> and <<router, routers>>
+Starting with _version 5.0_, the `setNotPropagatedHeaders()` implementation on the `AbstractMessageProducingHandler` applies the simple patterns (`xxx*`, `*xxx`, `*xxx*` or `xxx*yyy`) to allow to filter header with common suffix or prefix.
```
`s/to allow to filter with common/to allow filtering headers with a common/`

artembilan committed:
* Fix `message.adoc`
* Add `noHeadersPropagation` flag to the `AbstractMessageProducingHandler`
* Rework logic in the `updateNotPropagatedHeaders()` to store the array
of patterns instead of `Set` to avoid extra operation on each message
* Combine `noHeadersPropagation` with the `shouldCopyRequestHeaders()`
in the `createOutputMessage()` for logic to determine if we should start
the copy-headers procedure at all

artembilan commented:
OK. Pushed some fix.
I think ready for one more review round.
Thanks

artembilan committed:
* Revert `AbstractMessageProducingHandler.selectiveHeaderPropagation`
* Optimize `AmqpOutboundGatewayParserTests` performance from 3 secs to 0.5

artembilan committed:
Fix more NPEs in the `AbstractMessageProducingHandler`

garyrussell commented:
Merged as 90c46f5a79894a130badad77d2ee7ede9b33f697
